File: paobar_fluter/lib/features/now_playing/presentation/widgets/tab_renderer/tab_document.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'tab_document.freezed.dart';
part 'tab_document.g.dart';

/// 规范化吉他谱文档。字段完全对齐 `paobar_java` 的 `com.sosocom.tabdoc.TabDocument`，
/// 直接 `JSON ↔ freezed` 互通。爬虫入库时由 Java 侧 `TabHtmlNormalizer` 生成。
@freezed
class TabDocument with _$TabDocument {
  const factory TabDocument({
    @Default(1) int schemaVersion,
    String? title,
    @Default(<InfoItem>[]) List<InfoItem> info,
    String? meter,
    String? bpm,
    String? capoKey,
    String? originalKey,
    String? chordStyle,
    String? instrument,
    @Default(<SheetBlock>[]) List<SheetBlock> blocks,
  }) = _TabDocument;

  factory TabDocument.fromJson(Map<String, dynamic> json) =>
      _$TabDocumentFromJson(json);
}

/// "唱: 李健"、"编: 菜鸟吉他" 这样的 label-text 项。
@freezed
class InfoItem with _$InfoItem {
  const factory InfoItem({
    required String label,
    required String text,
  }) = _InfoItem;

  factory InfoItem.fromJson(Map<String, dynamic> json) =>
      _$InfoItemFromJson(json);
}

/// 正文段落块。union 辨别符使用 `type` 字段（值 "headline"/"paragraph"/"blank"）。
@Freezed(unionKey: 'type', unionValueCase: FreezedUnionCase.none)
sealed class SheetBlock with _$SheetBlock {
  const factory SheetBlock.headline({required String text}) = Headline;

  const factory SheetBlock.paragraph({
    @Default(<LineSegment>[]) List<LineSegment> segments,
  }) = Paragraph;

  const factory SheetBlock.blank() = Blank;

  factory SheetBlock.fromJson(Map<String, dynamic> json) =>
      _$SheetBlockFromJson(json);
}

/// 行内片段。union 辨别符同样为 `type`（值 "text"/"chord"）。
@Freezed(unionKey: 'type', unionValueCase: FreezedUnionCase.none)
sealed class LineSegment with _$LineSegment {
  /// 纯歌词。
  const factory LineSegment.text({required String text}) = PlainText;

  /// 某字带上方和弦名（chord 可为空串，表示纯占位锚点）。
  const factory LineSegment.chord({
    required String chord,
    @Default('') String text,
  }) = ChordText;

  factory LineSegment.fromJson(Map<String, dynamic> json) =>
      _$LineSegmentFromJson(json);
}

File: paobar_java/src/main/java/com/sosocom/tabdoc/TabDocument.java
package com.sosocom.tabdoc;

import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.Data;

import java.util.List;

/**
 * 规范化吉他谱文档 —— 爬虫入库时从源站 HTML 解析出的结构化表示。
 *
 * 设计要点：
 * - schemaVersion 用于未来字段演进时做兼容。
 * - blocks 是段落序列（headline / paragraph / blank）。
 * - 与 paobar_fluter 的 TabDocument 一一对应，直接 JSON 互通。
 */
@Data
@JsonInclude(JsonInclude.Include.NON_NULL)
public class TabDocument {

    /** 数据结构版本号，从 1 开始。 */
    private int schemaVersion = 1;

    /** 歌名。 */
    private String title;

    /** 演唱者 / 编谱 等信息条目。 */
    private List<InfoItem> info;

    /** 拍号，例如 "4/4"。 */
    private String meter;

    /** 拍速（BPM），字符串形式保留源站原样。 */
    private String bpm;

    /** 选调（演奏调），例如 "G"。 */
    private String capoKey;

    /** 原唱调，例如 "A"。 */
    private String originalKey;

    /** 和弦风格："number"（数字谱）| "letter"（字母谱）。 */
    private String chordStyle;

    /** 乐器："guitar" | "ukulele" | ...。 */
    private String instrument;

    /** 正文段落序列。 */
    private List<SheetBlock> blocks;
}
